### PR TITLE
Dc efs 485 0

### PR DESCRIPTION
**Pull request by drcate**
Re-worked presave with dates. Note that some code is duplicated on purpose, to separate date and string processing

### CODE REVIEW

File: src/preSaveHandlers/handlers/dateColumnHandler.js [hunk 1]
```diff
@@ -13,17 +30,79 @@ class DateColumnHandler extends PreSaveHandler {
...
+                if (Array.isArray(value)) {
+                    value.forEach((item, index) => {
+                    if (typeof item === 'object' && item !== null) {
+                        recurseData(item, extendPath(currentPath, index));
+                    } else {
+                        if (this.shouldUpdate(resource, extendPath(currentPath, index))) {
+                            object[key][index] = this.setDate(item);
+                        }
+                    }
+                });
```
Comment (007sunny) [inline src/preSaveHandlers/handlers/dateColumnHandler.js:57]: Can we replace this with recurseData(value, currentPath)? I think that might work the same, also simplifying the code

File: src/operations/common/resourceMerger.js [hunk 2]
```diff
@@ -7,9 +7,10 @@ const { PreSaveManager } = require('../../preSaveHandlers/preSave');
 const { assertTypeEquals, assertIsValid } = require('../../utils/assertType');
 const { getFirstElementOrNull } = require('../../utils/list.util');
 const { isUuid } = require('../../utils/uid.util');
-const { mergeObject } = require('../../utils/mergeHelper');
+let { mergeObject } = require('../../utils/mergeHelper');
```
Comment (007sunny) [inline src/operations/common/resourceMerger.js:10]: I think this change was by mistake, please have a look

File: src/tests/operations/query/r4/r4.test.js [hunk 3]
```diff
@@ -892,64 +892,103 @@ describe('r4 search Tests', () => {
...
+                                  }
                                 },
                                 {
-                                    effectiveInstant: {
-                                        $options: 'i',
-                                        $regex: /^(?:2019-10-16T22:12)|(?:2019-10-16T22:12:29)|(?:2019$)|(?:2019-10$)|(?:2019-10-16$)|(?:2019-10-16T22:12Z?$)/
-                                    }
+                                  'effectivePeriod.end': {
+                                    $type: 'string'
+                                  }
```
Comment (007sunny) [inline src/tests/operations/query/r4/r4.test.js:932]: Can we combine $type and $gte, $lte checks? Just like we did here https://github.com/icanbwell/fhir-server/pull/1635/files#diff-c011e3d57fc12bb2b1b701c9958c16f25e649228eff2c05eff23967fa6a261b1R964-R970

Also, we should not change the queries for fields that were already date-type as these fields worked as expected. Like meta.lastUpdated in all resources and recorded field AuditEvent.

File: src/tests/accessIndex/accessIndex/accessindex.test.js [hunk 4]
```diff
@@ -4,7 +4,9 @@ const patient1Resource = require('./fixtures/Patient/patient1.json');
 
 // expected
 const expectedAuditEventResources = require('./fixtures/expected/expected_AuditEvent.json');
+const expectedAuditEventResourcesThedcare = require('./fixtures/expected/expected_AuditEvent_thedcare.json');
 const expectedAuditEventWithoutAccessIndexResources = require('./fixtures/expected/expected_AuditEvent_without_access_index.json');
+const expectedAuditEventWithoutAccessIndexResourcesThedcare = require('./fixtures/expected/expected_AuditEvent_thedcare.json');
 const expectedAuditEventResourcesAccessIndex = require('./fixtures/expected/expected_AuditEvent_access_index.json');
```
Comment (007sunny) [inline src/tests/accessIndex/accessIndex/accessindex.test.js:10]: Can we remove unused imports and also delete their files?

Comment (drcate) [reply src/tests/accessIndex/accessIndex/accessindex.test.js:10]: Removed unused import on line 9. Line 7 is used in the test code.

File: src/tests/operations/query/r4/r4.test.js (hunk 3 above)
Comment (drcate) [reply src/tests/operations/query/r4/r4.test.js:932]: Updated code to combine checks. Date-type queries only changed to add $type: 'date', to match the 'string' logic.

File: src/tests/accessIndex/accessIndex/accessindex.test.js (hunk 4 above)
Comment (drcate) [reply src/tests/accessIndex/accessIndex/accessindex.test.js:10]: Removed unused import.

File: src/tests/operations/query/r4/r4.test.js (hunk 3 above)
Comment (007sunny) [reply src/tests/operations/query/r4/r4.test.js]:
```query before: {'recorded':{'$gt':ISODate('2021-04-02T00:00:00.000Z'),'$lt':ISODate('2021-09-02T00:00:00.000Z')}}
query after: {'$and':[{'recorded':{'$gt':ISODate('2021-04-02T00:00:00.000Z'),'$type':'date'}},{'recorded':{'$lt':ISODate('2021-09-02T00:00:00.000Z'),'$type':'date'}}]}```

We can see that $type: 'date' is added but $lt and $gt checks are separated into 2 queries. Can we avoid this?

File: src/preSaveHandlers/handlers/dateColumnHandler.js (hunk 1 above)
Comment (drcate) [reply src/preSaveHandlers/handlers/dateColumnHandler.js:57]: Tried switching that out, and dates in Arrays are not addressed, so I think we should leave as is.

File: src/tests/operations/query/r4/r4.test.js (hunk 3 above)
Comment (drcate) [reply src/tests/operations/query/r4/r4.test.js]: Updated code to combine $lt and $gt checks for this

**Review by 007sunny — APPROVED** [review]

**Pull request merged**